File: OsmAnd-java/src/main/java/net/osmand/binary/BinaryMapDataObject.java
package net.osmand.binary;

import gnu.trove.iterator.TIntObjectIterator;
import gnu.trove.list.array.TIntArrayList;
import gnu.trove.map.hash.TIntObjectHashMap;

import java.util.Arrays;
import java.util.LinkedHashMap;
import java.util.Map;

import net.osmand.binary.BinaryMapIndexReader.MapIndex;
import net.osmand.binary.BinaryMapIndexReader.TagValuePair;
import net.osmand.render.RenderingRulesStorage;
import net.osmand.util.Algorithms;

public class BinaryMapDataObject {
	protected int[] coordinates = null;
	protected int[][] polygonInnerCoordinates = null;
	protected boolean area = false;
	protected int[] types = null;
	protected int[] additionalTypes = null;
	protected int objectType = RenderingRulesStorage.POINT_RULES;
	protected int labelX;
	protected int labelY;
	private static final int SHIFT_ID = 7;
	
	protected TIntObjectHashMap<String> objectNames = null;
	protected TIntArrayList namesOrder = null;
	protected long id = 0;
	
	protected MapIndex mapIndex = null;

	public BinaryMapDataObject() {
	}

	public BinaryMapDataObject(long id, int[] coordinates, int[][] polygonInnerCoordinates, int objectType, boolean area, 
			int[] types, int[] additionalTypes, int labelX, int labelY) {
		this.polygonInnerCoordinates = polygonInnerCoordinates;
		this.coordinates = coordinates;
		this.additionalTypes = additionalTypes;
		this.types = types;
		this.id = id;
		this.objectType = objectType;
		this.area = area;
		this.labelX = labelX;
		this.labelY = labelY;
	}
	
	protected void setCoordinates(int[] coordinates) {
		this.coordinates = coordinates;
	}
	
	
	public String getName() {
		if(objectNames == null){
			return "";
		}
		String name = objectNames.get(mapIndex.nameEncodingType);
		if(name == null){
			return "";
		}
		return name;
	}

	public TIntObjectHashMap<String> getObjectNames() {
		return objectNames;
	}
	
	public Map<Integer, String> getOrderedObjectNames() {
		if (namesOrder == null) {
			return null;
		}
		LinkedHashMap<Integer, String> lm = new LinkedHashMap<Integer, String> ();
		for (int i = 0; i < namesOrder.size(); i++) {
			int nm = namesOrder.get(i);
			lm.put(nm, objectNames.get(nm));
		}
		return lm;
	}
	
	public void putObjectName(int type, String name){
		if(objectNames == null){
			objectNames = new TIntObjectHashMap<String>();
			namesOrder = new TIntArrayList();
		}
		objectNames.put(type, name);
		namesOrder.add(type);
	}
	
	public int[][] getPolygonInnerCoordinates() {
		return polygonInnerCoordinates;
	}
	
	public int[] getTypes(){
		return types;
	}
	
	public boolean containsType(int cachedType) {
		if(cachedType != -1) {
			for(int i=0; i<types.length; i++){
				if(types[i] == cachedType) {
					return true;
				}
			}
		}
		return false;
	}
	
	public boolean containsAdditionalType(int cachedType) {
		if (cachedType != -1 && additionalTypes != null) {
			for (int i = 0; i < additionalTypes.length; i++) {
				if (additionalTypes[i] == cachedType) {
					return true;
				}
			}
		}
		return false;
	}
	
	public String getNameByType(int type) {
		if(type != -1 && objectNames != null) {
			return objectNames.get(type);
		}
		return null;
	}
	
	public int[] getAdditionalTypes() {
		return additionalTypes;
	}
	
	public boolean isArea() {
		return area;
	}
	
	public boolean isCycle() {
		if(coordinates == null || coordinates.length < 2) {
			return false;
		}
		return coordinates[0] == coordinates[coordinates.length - 2] && 
				coordinates[1] == coordinates[coordinates.length - 1];
	}
	
	public void setArea(boolean area) {
		this.area = area;
	}
	
	public long getId() {
		return id;
	}
	
	protected void setId(long id) {
		this.id = id;
	}
	
	protected void setTypes(int[] types) {
		this.types = types;
	}
	
	
	public int getSimpleLayer() {
		if (mapIndex != null) {
			if (additionalTypes != null) {
				for (int i = 0; i < additionalTypes.length; i++) {
					if (mapIndex.positiveLayers.contains(additionalTypes[i])) {
						return 1;
					} else if (mapIndex.negativeLayers.contains(additionalTypes[i])) {
						return -1;
					}
				}
			}
		}
		return 0;
	}
	
	public TIntArrayList getNamesOrder() {
		return namesOrder;
	}
	
	public MapIndex getMapIndex() {
		return mapIndex;
	}
	
	public void setMapIndex(MapIndex mapIndex) {
		this.mapIndex = mapIndex;
	}
	
	public int getPointsLength() {
		if(coordinates == null){
			return 0;
		}
		return coordinates.length / 2;
	}

	public int getPoint31YTile(int ind) {
		return coordinates[2 * ind + 1];
	}

	public int getPoint31XTile(int ind) {
		return coordinates[2 * ind];
	}
	
	public boolean compareBinary(BinaryMapDataObject thatObj, int coordinatesPrecision) {
		if(this.objectType == thatObj.objectType
				&& this.id == thatObj.id
				&& this.area == thatObj.area 
				&& compareCoordinates(this.coordinates, thatObj.coordinates, coordinatesPrecision) ) {
			if (mapIndex == null) {
				throw new IllegalStateException("Illegal binary object: " + id);
			}
			if (thatObj.mapIndex == null) {
				throw new IllegalStateException("Illegal binary object: " + thatObj.id);
			}
			
			boolean equals = true;
			if (equals) {
				if (polygonInnerCoordinates == null || thatObj.polygonInnerCoordinates == null) {
					equals = polygonInnerCoordinates == thatObj.polygonInnerCoordinates;
				} else if (polygonInnerCoordinates.length != thatObj.polygonInnerCoordinates.length) {
					equals = false;
				} else {
					for (int i = 0; i < polygonInnerCoordinates.length && equals; i++) {
						if (polygonInnerCoordinates[i] == null || thatObj.polygonInnerCoordinates[i] == null) {
							equals = polygonInnerCoordinates[i] == thatObj.polygonInnerCoordinates[i];
						} else if (polygonInnerCoordinates[i].length != thatObj.polygonInnerCoordinates[i].length) {
							equals = false;
						} else {
							equals = compareCoordinates(polygonInnerCoordinates[i], thatObj.polygonInnerCoordinates[i],
									coordinatesPrecision);
						}
					}
				}
			}
			
			if (equals) {
				if (types == null || thatObj.types == null) {
					equals = types == thatObj.types;
				} else if (types.length != thatObj.types.length) {
					equals = false;
				} else {
					for (int i = 0; i < types.length && equals; i++) {
						TagValuePair o = mapIndex.decodeType(types[i]);
						TagValuePair s = thatObj.mapIndex.decodeType(thatObj.types[i]);
						equals = o.equals(s) && equals;
					}
				}
			}
			if (equals) {
				if (additionalTypes == null || thatObj.additionalTypes == null) {
					equals = additionalTypes == thatObj.additionalTypes;
				} else if (additionalTypes.length != thatObj.additionalTypes.length) {
					equals = false;
				} else {
					for (int i = 0; i < additionalTypes.length && equals; i++) {
						TagValuePair o = mapIndex.decodeType(additionalTypes[i]);
						TagValuePair s = thatObj.mapIndex.decodeType(thatObj.additionalTypes[i]);
						equals = o.equals(s);
					}
				}
			}
			if (equals) {
				if (namesOrder == null || thatObj.namesOrder == null) {
					equals = namesOrder == thatObj.namesOrder;
				} else if (namesOrder.size() != thatObj.namesOrder.size()) {
					equals = false;
				} else {
					for (int i = 0; i < namesOrder.size() && equals; i++) {
						TagValuePair o = mapIndex.decodeType(namesOrder.get(i));
						TagValuePair s = thatObj.mapIndex.decodeType(thatObj.namesOrder.get(i));
						equals = o.equals(s);
					}
				}
			}
			if (equals) {
				// here we know that name indexes are equal & it is enough to check the value sets
				if (objectNames == null || thatObj.objectNames == null) {
					equals = objectNames == thatObj.objectNames;
				} else if (objectNames.size() != thatObj.objectNames.size()) {
					equals = false;
				} else {
					for (int i = 0; i < namesOrder.size() && equals; i++) {
						String o = objectNames.get(namesOrder.get(i));
						String s = thatObj.objectNames.get(thatObj.namesOrder.get(i));
						equals = Algorithms.objectEquals(o, s);
					}
				}
			}
			
			return equals;
		}
//		thatObj.mapIndex.decodeType(thatObj.types[0])
//		mapIndex.decodeType(types[0]) id >>7
		return false;
	}


	private static boolean compareCoordinates(int[] coordinates, int[] coordinates2, int precision) {
		if(precision == 0) {
			return Arrays.equals(coordinates, coordinates2);
		}
		TIntArrayList cd = simplify(coordinates, precision);
		TIntArrayList cd2 = simplify(coordinates2, precision);
		return cd.equals(cd2);
	}



	private static TIntArrayList simplify(int[] c, int precision) {
		int len = c.length / 2;
		TIntArrayList lt = new TIntArrayList(len * 3);
		for (int i = 0; i < len; i++) {
			lt.add(0);
			lt.add(c[i * 2]);
			lt.add(c[i * 2 + 1]);
		}
		lt.set(0, 1);
		lt.set((len - 1) * 3, 1);
		simplifyLine(lt, precision, 0, len - 1);

		TIntArrayList res = new TIntArrayList(len * 2);
		for (int i = 0; i < len; i++) {
			if (lt.get(i * 3) == 1) {
				res.add(lt.get(i * 3 + 1));
				res.add(lt.get(i * 3 + 2));
			}
		}
		return res;
	}

	private static double orthogonalDistance(int x, int y, int x1, int y1, int x2, int y2) {
		long A = (x - x1);
		long B = (y - y1);
		long C = (x2 - x1);
		long D = (y2 - y1);
		return Math.abs(A * D - C * B) / Math.sqrt(C * C + D * D);
	}

	private static void simplifyLine(TIntArrayList lt, int precision, int start, int end) {
		if(start == end - 1) {
			return;
		}
		int x = lt.get(start*3 + 1);
		int y = lt.get(start*3 + 2);
		int ex = lt.get(end*3 + 1);
		int ey = lt.get(end*3 + 2);
		double max = 0;
		int maxK = -1;
		for(int k = start + 1; k < end ; k++) {
			double ld = orthogonalDistance(lt.get(k*3 + 1), lt.get(k*3 + 2), x, y, ex, ey);
			if(maxK == -1 || max < ld) {
				maxK = k;
				max = ld;
			}
		}
		if(max < precision) {
			return;
		}
		lt.set(maxK*3, 1); // keep point
		simplifyLine(lt, precision, start, maxK);
		simplifyLine(lt, precision, maxK, end);
		
	}

	public boolean isLabelSpecified() {
		return (labelX != 0 || labelY != 0) && coordinates.length > 0;
	}

	public int getLabelX() {
		long sum = 0;
		int LABEL_SHIFT = 31 - BinaryMapIndexReader.LABEL_ZOOM_ENCODE;
		int len = coordinates.length / 2;
		for(int i = 0; i < len; i++) {
			sum += coordinates[2 * i];
		}
		int average = ((int) (sum >> BinaryMapIndexReader.SHIFT_COORDINATES)/ len) 
				<< (BinaryMapIndexReader.SHIFT_COORDINATES - LABEL_SHIFT);
		int label31X = (average + this.labelX) << LABEL_SHIFT;
		return label31X;
	}
	
	public int getLabelY() {
		long sum = 0;
		int LABEL_SHIFT = 31 - BinaryMapIndexReader.LABEL_ZOOM_ENCODE;
		int len = coordinates.length / 2;
		for(int i = 0; i < len; i++) {
			sum += coordinates[2 * i + 1];
		}
		int average = ((int) (sum >> BinaryMapIndexReader.SHIFT_COORDINATES)/ len) 
				<< (BinaryMapIndexReader.SHIFT_COORDINATES - LABEL_SHIFT);
		int label31Y = (average + this.labelY) << LABEL_SHIFT;
		return label31Y;
	}

	public int[] getCoordinates() {
		return coordinates;
	}
	
	public int getObjectType() {
		return objectType;
	}

	public String getTagValue(String tag) {
		if (mapIndex == null) {
			return "";
		}
		TIntObjectIterator<String> it = objectNames.iterator();
		while (it.hasNext()) {
			it.advance();
			BinaryMapIndexReader.TagValuePair tp = mapIndex.decodeType(it.key());
			if (tp.tag.equals(tag)) {
				return it.value();
			}
		}
		return "";
	}

	public String getAdditionalTagValue(String tag) {
		if (mapIndex == null) {
			return "";
		}
		for (int type : additionalTypes) {
			TagValuePair tp = mapIndex.decodeType(type);
			if (tag.equals(tp.tag)) {
				return tp.value;
			}
		}
		return "";
	}
	
	@Override
	public String toString() {
		String obj = "Point";
		if(objectType == RenderingRulesStorage.LINE_RULES) {
			obj = "Line";
		} else if(objectType == RenderingRulesStorage.POLYGON_RULES) {
			obj = "Polygon";
		}
		return obj + " " + (getId() >> SHIFT_ID);
	}
}

File: OsmAnd-shared/src/commonMain/kotlin/net/osmand/shared/binary/BinaryMapDataObject.kt
package net.osmand.shared.binary

import net.osmand.shared.data.KLatLon
import net.osmand.shared.util.KMapUtils
import net.osmand.shared.util.collections.KTIntArrayList
import net.osmand.shared.util.collections.KTIntObjectMap
import kotlin.jvm.JvmField
import kotlin.math.abs
import kotlin.math.sqrt

/**
 * One object of a map section: its points in 31 coordinates, the numbers of its tags, and its
 * names. The numbers mean nothing without [mapIndex], which holds the table that decodes them.
 *
 * A copy of `BinaryMapDataObject` in OsmAnd-java, which stays there for android and tools; this
 * copy is for iOS. Two differences from java. `toString` prints the raw id, because turning it
 * back into an osm id belongs to `ObfConstants`, which is not copied yet. And where java hands
 * back a null it read - a name the string table had nothing for, a type its section has no rule
 * for - this answers "" or leaves the entry out instead of throwing; the reader fills names and
 * their order together, so no object read out of a file gets there.
 */
class BinaryMapDataObject {

	@JvmField
	var coordinates: IntArray? = null

	@JvmField
	var polygonInnerCoordinates: Array<IntArray>? = null

	@JvmField
	var area: Boolean = false

	@JvmField
	var types: IntArray? = null

	@JvmField
	var additionalTypes: IntArray? = null

	@JvmField
	var objectType: Int = POINT_RULES

	@JvmField
	var labelX: Int = 0

	@JvmField
	var labelY: Int = 0

	@JvmField
	var objectNames: KTIntObjectMap<String>? = null

	@JvmField
	var namesOrder: KTIntArrayList? = null

	@JvmField
	var id: Long = 0

	@JvmField
	var mapIndex: MapIndex? = null

	constructor()

	constructor(
		id: Long, coordinates: IntArray?, polygonInnerCoordinates: Array<IntArray>?,
		objectType: Int, area: Boolean, types: IntArray?, additionalTypes: IntArray?,
		labelX: Int, labelY: Int
	) {
		this.polygonInnerCoordinates = polygonInnerCoordinates
		this.coordinates = coordinates
		this.additionalTypes = additionalTypes
		this.types = types
		this.id = id
		this.objectType = objectType
		this.area = area
		this.labelX = labelX
		this.labelY = labelY
	}

	fun setCoordinates(coordinates: IntArray?) {
		this.coordinates = coordinates
	}

	fun getName(): String {
		val names = objectNames ?: return ""
		return names[mapIndex?.nameEncodingType ?: 0] ?: ""
	}

	fun getObjectNames(): KTIntObjectMap<String>? = objectNames

	/** The names in the order the file lists them, which is the order they should be shown in. */
	fun getOrderedObjectNames(): Map<Int, String>? {
		val order = namesOrder ?: return null
		val lm = LinkedHashMap<Int, String>()
		for (i in 0 until order.size()) {
			val nm = order[i]
			val name = objectNames?.get(nm)
			if (name != null) {
				lm[nm] = name
			}
		}
		return lm
	}

	fun putObjectName(type: Int, name: String) {
		if (objectNames == null) {
			objectNames = KTIntObjectMap()
			namesOrder = KTIntArrayList()
		}
		objectNames!!.put(type, name)
		namesOrder!!.add(type)
	}

	fun getPolygonInnerCoordinates(): Array<IntArray>? = polygonInnerCoordinates

	fun getTypes(): IntArray? = types

	fun containsType(cachedType: Int): Boolean {
		val types = this.types
		if (cachedType != -1 && types != null) {
			for (type in types) {
				if (type == cachedType) {
					return true
				}
			}
		}
		return false
	}

	fun containsAdditionalType(cachedType: Int): Boolean {
		val additionalTypes = this.additionalTypes
		if (cachedType != -1 && additionalTypes != null) {
			for (type in additionalTypes) {
				if (type == cachedType) {
					return true
				}
			}
		}
		return false
	}

	fun getNameByType(type: Int): String? {
		if (type != -1) {
			return objectNames?.get(type)
		}
		return null
	}

	fun getAdditionalTypes(): IntArray? = additionalTypes

	fun isArea(): Boolean = area

	fun isCycle(): Boolean {
		val coordinates = this.coordinates
		if (coordinates == null || coordinates.size < 2) {
			return false
		}
		return coordinates[0] == coordinates[coordinates.size - 2] &&
				coordinates[1] == coordinates[coordinates.size - 1]
	}

	fun setArea(area: Boolean) {
		this.area = area
	}

	fun getId(): Long = id

	fun setId(id: Long) {
		this.id = id
	}

	fun setTypes(types: IntArray?) {
		this.types = types
	}

	/** 1 for a bridge, -1 for a tunnel, 0 for anything at ground level. */
	fun getSimpleLayer(): Int {
		val mapIndex = this.mapIndex
		val additionalTypes = this.additionalTypes
		if (mapIndex != null && additionalTypes != null) {
			for (type in additionalTypes) {
				if (mapIndex.positiveLayers.contains(type)) {
					return 1
				} else if (mapIndex.negativeLayers.contains(type)) {
					return -1
				}
			}
		}
		return 0
	}

	fun getNamesOrder(): KTIntArrayList? = namesOrder

	fun getMapIndex(): MapIndex? = mapIndex

	fun setMapIndex(mapIndex: MapIndex?) {
		this.mapIndex = mapIndex
	}

	fun getPointsLength(): Int {
		val coordinates = this.coordinates ?: return 0
		return coordinates.size / 2
	}

	fun getPoint31YTile(ind: Int): Int = coordinates!![2 * ind + 1]

	fun getPoint31XTile(ind: Int): Int = coordinates!![2 * ind]

	/**
	 * Whether [thatObj] is the same object, read from another file: same id and shape, and the same
	 * tags once both sides are decoded through their own sections. A [coordinatesPrecision] above
	 * zero allows the two shapes to differ by that much, for maps generated at different detail.
	 */
	fun compareBinary(thatObj: BinaryMapDataObject, coordinatesPrecision: Int): Boolean {
		if (this.objectType == thatObj.objectType &&
			this.id == thatObj.id &&
			this.area == thatObj.area &&
			compareCoordinates(this.coordinates, thatObj.coordinates, coordinatesPrecision)
		) {
			val mapIndex = this.mapIndex ?: throw IllegalStateException("Illegal binary object: $id")
			val thatMapIndex = thatObj.mapIndex
				?: throw IllegalStateException("Illegal binary object: ${thatObj.id}")

			var equals = true
			if (equals) {
				val inner = polygonInnerCoordinates
				val thatInner = thatObj.polygonInnerCoordinates
				if (inner == null || thatInner == null) {
					equals = inner === thatInner
				} else if (inner.size != thatInner.size) {
					equals = false
				} else {
					var i = 0
					while (i < inner.size && equals) {
						equals = compareCoordinates(inner[i], thatInner[i], coordinatesPrecision)
						i++
					}
				}
			}

			if (equals) {
				val types = this.types
				val thatTypes = thatObj.types
				if (types == null || thatTypes == null) {
					equals = types === thatTypes
				} else if (types.size != thatTypes.size) {
					equals = false
				} else {
					var i = 0
					while (i < types.size && equals) {
						val o = mapIndex.decodeType(types[i])
						val s = thatMapIndex.decodeType(thatTypes[i])
						equals = o == s && equals
						i++
					}
				}
			}
			if (equals) {
				val additionalTypes = this.additionalTypes
				val thatAdditionalTypes = thatObj.additionalTypes
				if (additionalTypes == null || thatAdditionalTypes == null) {
					equals = additionalTypes === thatAdditionalTypes
				} else if (additionalTypes.size != thatAdditionalTypes.size) {
					equals = false
				} else {
					var i = 0
					while (i < additionalTypes.size && equals) {
						val o = mapIndex.decodeType(additionalTypes[i])
						val s = thatMapIndex.decodeType(thatAdditionalTypes[i])
						equals = o == s
						i++
					}
				}
			}
			if (equals) {
				val namesOrder = this.namesOrder
				val thatNamesOrder = thatObj.namesOrder
				if (namesOrder == null || thatNamesOrder == null) {
					equals = namesOrder === thatNamesOrder
				} else if (namesOrder.size() != thatNamesOrder.size()) {
					equals = false
				} else {
					var i = 0
					while (i < namesOrder.size() && equals) {
						val o = mapIndex.decodeType(namesOrder[i])
						val s = thatMapIndex.decodeType(thatNamesOrder[i])
						equals = o == s
						i++
					}
				}
			}
			if (equals) {
				// here we know that name indexes are equal & it is enough to check the value sets
				val objectNames = this.objectNames
				val thatObjectNames = thatObj.objectNames
				if (objectNames == null || thatObjectNames == null) {
					equals = objectNames === thatObjectNames
				} else if (objectNames.size() != thatObjectNames.size()) {
					equals = false
				} else {
					val namesOrder = this.namesOrder
					val thatNamesOrder = thatObj.namesOrder
					var i = 0
					while (namesOrder != null && thatNamesOrder != null && i < namesOrder.size() && equals) {
						val o = objectNames[namesOrder[i]]
						val s = thatObjectNames[thatNamesOrder[i]]
						equals = o == s
						i++
					}
				}
			}

			return equals
		}
		return false
	}

	fun isLabelSpecified(): Boolean =
		(labelX != 0 || labelY != 0) && (coordinates?.size ?: 0) > 0

	fun getLabelX(): Int {
		var sum = 0L
		val coordinates = this.coordinates!!
		val labelShift = 31 - BinaryMapIndexReader.LABEL_ZOOM_ENCODE
		val len = coordinates.size / 2
		for (i in 0 until len) {
			sum += coordinates[2 * i]
		}
		val l = ((sum shr BinaryMapIndexReader.SHIFT_COORDINATES) / len).toInt()
		val average = l shl (BinaryMapIndexReader.SHIFT_COORDINATES - labelShift)
		return (average + this.labelX) shl labelShift
	}

	fun getLabelY(): Int {
		var sum = 0L
		val coordinates = this.coordinates!!
		val labelShift = 31 - BinaryMapIndexReader.LABEL_ZOOM_ENCODE
		val len = coordinates.size / 2
		for (i in 0 until len) {
			sum += coordinates[2 * i + 1]
		}
		val l = ((sum shr BinaryMapIndexReader.SHIFT_COORDINATES) / len).toInt()
		val average = l shl (BinaryMapIndexReader.SHIFT_COORDINATES - labelShift)
		return (average + this.labelY) shl labelShift
	}

	fun getLabelLatLon(): KLatLon =
		KLatLon(KMapUtils.get31LatitudeY(getLabelY()), KMapUtils.get31LongitudeX(getLabelX()))

	fun getCoordinates(): IntArray? = coordinates

	fun getObjectType(): Int = objectType

	fun getTagValue(tag: String): String {
		val mapIndex = this.mapIndex ?: return ""
		val names = objectNames ?: return ""
		for (key in names.keys()) {
			val tp = mapIndex.decodeType(key)
			if (tp != null && tp.tag == tag) {
				return names[key] ?: ""
			}
		}
		return ""
	}

	fun getAdditionalTagValue(tag: String): String {
		val mapIndex = this.mapIndex ?: return ""
		val additionalTypes = this.additionalTypes ?: return ""
		for (type in additionalTypes) {
			val tp = mapIndex.decodeType(type)
			if (tp != null && tag == tp.tag) {
				return tp.value ?: ""
			}
		}
		return ""
	}

	override fun toString(): String {
		val obj = when (objectType) {
			LINE_RULES -> "Line"
			POLYGON_RULES -> "Polygon"
			else -> "Point"
		}
		return "$obj $id"
	}

	fun isDeleted(): Boolean {
		val mapIndex = this.mapIndex
		val types = this.types
		if (mapIndex != null && types != null && types.isNotEmpty()) {
			for (t in types) {
				val tvp = mapIndex.decodeType(t)
				if (tvp != null && tvp.tag == "osmand_change") {
					return true
				}
			}
		}
		return false
	}

	companion object {
		const val SHIFT_ID: Int = 7

		// the three kinds of rendering rule an object can be drawn by, as RenderingRulesStorage
		// in OsmAnd-java numbers them
		const val POINT_RULES: Int = 1
		const val LINE_RULES: Int = 2
		const val POLYGON_RULES: Int = 3

		private fun compareCoordinates(
			coordinates: IntArray?, coordinates2: IntArray?, precision: Int
		): Boolean {
			if (precision == 0) {
				return coordinates.contentEquals(coordinates2)
			}
			if (coordinates == null || coordinates2 == null) {
				return coordinates === coordinates2
			}
			val cd = simplify(coordinates, precision)
			val cd2 = simplify(coordinates2, precision)
			return cd == cd2
		}

		/** Douglas-Peucker: the points that carry the shape at [precision], in order. */
		private fun simplify(c: IntArray, precision: Int): KTIntArrayList {
			val len = c.size / 2
			val lt = KTIntArrayList(len * 3)
			for (i in 0 until len) {
				lt.add(0)
				lt.add(c[i * 2])
				lt.add(c[i * 2 + 1])
			}
			lt[0] = 1
			lt[(len - 1) * 3] = 1
			simplifyLine(lt, precision, 0, len - 1)

			val res = KTIntArrayList(len * 2)
			for (i in 0 until len) {
				if (lt[i * 3] == 1) {
					res.add(lt[i * 3 + 1])
					res.add(lt[i * 3 + 2])
				}
			}
			return res
		}

		private fun orthogonalDistance(x: Int, y: Int, x1: Int, y1: Int, x2: Int, y2: Int): Double {
			val a = (x - x1).toLong()
			val b = (y - y1).toLong()
			val c = (x2 - x1).toLong()
			val d = (y2 - y1).toLong()
			return abs(a * d - c * b) / sqrt((c * c + d * d).toDouble())
		}

		private fun simplifyLine(lt: KTIntArrayList, precision: Int, start: Int, end: Int) {
			if (start == end - 1) {
				return
			}
			val x = lt[start * 3 + 1]
			val y = lt[start * 3 + 2]
			val ex = lt[end * 3 + 1]
			val ey = lt[end * 3 + 2]
			var maxDistance = 0.0
			var maxK = -1
			for (k in start + 1 until end) {
				val ld = orthogonalDistance(lt[k * 3 + 1], lt[k * 3 + 2], x, y, ex, ey)
				if (maxK == -1 || maxDistance < ld) {
					maxK = k
					maxDistance = ld
				}
			}
			if (maxDistance < precision) {
				return
			}
			lt[maxK * 3] = 1 // keep point
			simplifyLine(lt, precision, start, maxK)
			simplifyLine(lt, precision, maxK, end)
		}
	}
}
